File: jwt/src/main/java/com/example/jwt/config/auth/PrincipalDetailService.java
package com.example.jwt.config.auth;

import com.example.jwt.model.User;
import com.example.jwt.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.core.userdetails.UserDetailsService;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.stereotype.Service;


// http://localhost:8081/login 이 주소일떄 동작인데
// SecurityConfig 에서 .formLogin().disable() 를 해서 동작을 안함
@Service
@RequiredArgsConstructor
public class PrincipalDetailService implements UserDetailsService {

    private final UserRepository userRepository;

    // 스프링이 로그인 요청을 가로챌 때, username, password 변수 2개를 가로채는데
    // password 부분 처리는 알아서 함.
    // username 이 DB에 있는지만 확인해주면 됨
    @Override
    public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException {
        System.out.println("PrincipalDetailsService의 loadUserByUsername()");
        User principal = userRepository.findByUsername(username)
                .orElseThrow(()->{
                    return new UsernameNotFoundException("해당 사용자를 찾을 수 없습니다. : " + username);
                });
        return new PrincipalDetails(principal); // 시큐리티의 세션에 유저 정보가 저장이 됨
    }
}

File: kotlinWithSecurityJWT/src/main/kotlin/com/springpratice3370/kotlinWithSecurityJWT/security/PrincipalDetailService.kt
package com.springpratice3370.kotlinWithSecurityJWT.security

import com.springpratice3370.kotlinWithSecurityJWT.user.User
import com.springpratice3370.kotlinWithSecurityJWT.user.UserRepository
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.security.core.userdetails.UserDetailsService
import org.springframework.stereotype.Service

@Service
class PrincipalDetailService(
    private val userRepository: UserRepository
) : UserDetailsService {

    override fun loadUserByUsername(email: String): UserDetails {
        val principal: User = userRepository.findByEmail(email) ?: throw RuntimeException()
        return PrincipalDetails(principal)
    }
}